### PR TITLE
[CALCITE-3007] Type mismatch for ANY subquery in project (Vineet Garg)

### PR DESCRIPTION
**Pull request by vineetgarg02**

### CODE REVIEW

File: core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml [hunk 1]
```diff
@@ -493,12 +493,86 @@ LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$
...
+]]>
+        </Resource>
+        <Resource name="planAfter">
+            <![CDATA[
+LogicalProject(DEPTNO=[$0], EXPR$1=[OR(AND(IS TRUE(=($1, $2)), <>($3, 0)), AND(>($3, $4), null, <>($3, 0), IS NOT TRUE(=($1, $2))), AND(=($1, $2), <>($3, 0), IS NOT TRUE(=($1, $2)), <=($3, $4)))])
+  LogicalJoin(condition=[true], joinType=[inner])
+    LogicalTableScan(table=[[CATALOG, SALES, DEPT]])
+    LogicalAggregate(group=[{}], m=[MAX($0)], c=[COUNT()], d=[COUNT($0)])
+      LogicalProject(MGR=[$3])
+        LogicalTableScan(table=[[CATALOG, SALES, EMP]])
```
Comment (hsyuan) [inline core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml]: PR https://github.com/apache/calcite/pull/1161/ will fix this wrong plan, right?

Comment (vineetgarg02) [reply core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml]: @hsyuan That is correct

File: core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml [hunk 2]
```diff
@@ -7466,7 +7540,7 @@ LogicalProject(DEPTNO=[$0])
             <![CDATA[
 LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], DEPTNO=[$7], SLACKER=[$8])
   LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], DEPTNO=[$7], SLACKER=[$8])
-    LogicalFilter(condition=[OR(AND(>($0, $9), <>($10, 0)), AND(>($0, $9), <>($10, 0), IS NOT TRUE(>($0, $9)), <=($10, $11)))])
+    LogicalFilter(condition=[OR(AND(IS TRUE(>($0, $9)), <>($10, 0)), AND(>($10, $11), null, <>($10, 0), IS NOT TRUE(>($0, $9))), AND(>($0, $9), <>($10, 0), IS NOT TRUE(>($0, $9)), <=($10, $11)))])
```
Comment (hsyuan) [inline core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml]: Just notice this, any idea for this change?

Comment (vineetgarg02) [reply core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml]: @hsyuan This is the result of CASE simplification. With cast it looks like CASE was simplified to this. I went over this and it looks semantically correct to me.

**Comment by hsyuan** [comment]
Looks like this patch caused test failure: https://builds.apache.org/job/Calcite-Master/jdk=JDK%201.9%20(latest),label_exp=ubuntu&&!cloud-slave&&!H27/lastBuild/console

I don't know why.

**Comment by vineetgarg02** [comment]
> Looks like this patch caused test failure: https://builds.apache.org/job/Calcite-Master/jdk=JDK%201.9%20(latest),label_exp=ubuntu&&!cloud-slave&&!H27/lastBuild/console
> 
> I don't know why.

That is strange. I am unable to see this failure locally. May be some environmental/configuration issue?

**Pull request closed without merging**